### ISSUE TITLE
In "Navigate To" have smarter support for camelcase searching

### ISSUE THREAD

**Issue by kuhlenh** [dotnet/roslyn#9346]
Here is an example:

1) I'm looking for "IntroduceVariableResult", but I don't remember the full name. I do know that it follows the pattern "IVR"
2) I use `Ctrl+,` to activate Navigate To and I type in "IVR"
3) This is what I now see:
![image](https://cloud.githubusercontent.com/assets/1402694/13437839/24bf6c5c-df9b-11e5-9c34-c9be85e9c0d4.png)
4) After a fair bit of scrolling, I see "IIntroduceVariableResult" as the 25th element and I see my desired "IntroduceVariableResult" as the 35th element.

My ask:
Can we have exact camelcase matches appear first? So "IntroduceVariableResult" would appear first, "IIntroduceVariableResult" would be lower down, and things like "_lazyIsVbRuntime" will filter out near the bottom? Or just have some way of ordering the results in a "smarter" way?

This is somewhat related to #8924.

**Comment by Pilchie** [comment]
While we might be able to do something based on the sort text of the matched item, I'm not sure we want to.  It is nice to have items in a predictably sorted alphabetic order.

**Comment by eamodio** [comment]
Since the camelcase matching only works if you use all CAPS, I would argue that should be reflected in the rankings. Also to me, alphabetic sorting should be the fallback, when no better ranking could be provided, but certainly not the only ranking.

**Comment by CyrusNajmabadi** [comment]
Have started a concersation with the editor about this. The primary issue is that the editor forces sorting by name when you have the same type of match.  i.e. between two CamelCase matches, we will always get name sorting.

I'm discussing mechanisms with there where the LS can supply an override for this so we can dictate what we want the actual order ot be.

**Comment by CyrusNajmabadi** [comment]
Pushing out to 2.1.  The editor woudl like to do this.  But we want a coordinated piece of work whereby we move CamelCase+Fuzzy matching down to their level

**Comment by CyrusNajmabadi** [comment]
@olegtk We could accomplish this very quickly if we just augmented MatchKind to have the new additional CamelCaseXXX values we talked about.  You would have to update the comparer on your end to understand these.  But this would allow us to get these better sorts without needing a total API overhaul.

**Comment by sharwell** [comment]
💭 This may be fixed by #24191

**Comment by jinujoseph** [comment]
@chborl Can you pls validate this

**Comment by chborl** [comment]
Yes, after #24191, the exact camelcase matches appear before the prefix camelcase and non-contiguous matches.

**Comment by jinujoseph** [comment]
thanks @chborl , closing based on that
